### PR TITLE
Add conjugate_encoder.R, change README, and add associated basic tests

### LINKED ISSUE #28: Code conjugate_encoder
This is just an issue to track the progress of the entire function (since the milestone instructions state to create an issue for each function, and we only created issues for each docstring in particular).

- [x]  Docstring
- [x]  Code
1. Implement function with exceptions.

Will modify as we understand what we need to do in the coming weeks.

### LINKED ISSUE #38: Code test_conjugate encoder
- [ ]  Comments explaining each test.
- [ ]  2 to 3 edge case tests

### PR DESCRIPTION
**Pull request by braydentang1**
For issue #28 and #38.

Still working on the tests.

### CODE REVIEW

File: R/conjugate_encoder.R [hunk 1]
```diff
@@ -36,6 +36,8 @@ library(rlang)
 #' objective = "regression")
 conjugate_encoder <- function(X_train, X_test = NA, y, cat_columns, prior_params, objective = "regression") {
```
Comment (bbaillie) [inline R/conjugate_encoder.R]: Inconsistency with my code in that you used X_test=NA instead of X_test = NULL, not sure if changing it will mess up another part of your code. 
```suggestion
conjugate_encoder <- function(X_train, X_test = NULL, y, cat_columns, prior_params, objective = "regression") {
```

File: R/conjugate_encoder.R [hunk 2]
```diff
@@ -10,7 +10,7 @@ library(rlang)
 #' per each category are used as the encodings.
 #'
 #' @param X_train A tibble representing the training data set containing some categorical features/columns.
-#' @param X_test A tibble representing the test set, containing some set of categorical features/columns.
+#' @param X_test An optional tibble representing the test set, containing some set of categorical features/columns. Default = NA.
```
Comment (bbaillie) [inline R/conjugate_encoder.R]: Same difference of NA vs NULL, changing for consistency
```suggestion
#' @param X_test An optional tibble representing the test set, containing some set of categorical features/columns. Default = NULL.
```

File: R/conjugate_encoder.R [hunk 3]
```diff
@@ -104,6 +103,8 @@ conjugate_encoder <- function(X_train, X_test, y, cat_columns, prior_params, obj
       train_processed <- reduce(encodings_list, function(x, y) left_join(x, y, by = NULL)) %>%
         select(-all_of(cat_columns))
 
+      if (!is.na(X_test)) {
```
Comment (bbaillie) [inline R/conjugate_encoder.R]: for consistency
```suggestion
      if (!is.null(X_test)) {
```

File: R/conjugate_encoder.R [hunk 4]
```diff
@@ -176,6 +178,8 @@ conjugate_encoder <- function(X_train, X_test, y, cat_columns, prior_params, obj
...
```
Comment (bbaillie) [inline R/conjugate_encoder.R]: for consistency
```suggestion
      if (!is.null(X_test)) {
```

File: R/conjugate_encoder.R [hunk 5]
```diff
@@ -186,9 +190,17 @@ conjugate_encoder <- function(X_train, X_test, y, cat_columns, prior_params, obj
       test_processed <- bind_cols(X_test, test_encodings) %>%
         select(-all_of(cat_columns))
 
+      }
+
     }
 
+  if (!is.na(X_test)) {
```
Comment (bbaillie) [inline R/conjugate_encoder.R]: for consistency
```suggestion
  if (!is.null(X_test)) {
```

File: R/conjugate_encoder.R [hunk 6]
```diff
@@ -31,6 +36,159 @@
...
+        select(contains("encoded"))
+
+      test_encodings[is.na(test_encodings)] <- alpha / (alpha + beta)
+
+      test_processed <- bind_cols(X_test, test_encodings) %>%
+        select(-all_of(cat_columns))
+
+    }
+
   out <- list(train_processed, test_processed)
```
Comment (bbaillie) [inline R/conjugate_encoder.R]: I have this list with named values so maybe for consistency this should be changed 
```suggestion
  out <- list("train" = train_processed, "test" = test_processed)
```

File: R/conjugate_encoder.R (hunk 1 above)
Comment (braydentang1) [reply R/conjugate_encoder.R]: I agree! Don't know what I thinking there, typically it's always assigned NULL like you suggested. Thanks!

File: R/conjugate_encoder.R (hunk 6 above)
Comment (braydentang1) [reply R/conjugate_encoder.R]: Sure, sounds good.

**Comment by braydentang1** [comment]
@bbaillie The issues have been addressed, and I have added the tests. Providing you approve, this can be merged.

**Review by bbaillie — COMMENTED** [review]
Thanks for implementing my suggested changes, everything looks great now

**Review by bbaillie — APPROVED** [review]
